api/openapi-spec.yaml

```diff
@@ -0,0 +1,349 @@
+        200:
+          description: Successfully retrieved the events
+          content:
+            application/json:
+              schema:
+                type: object
+                required:
+                    - pageSize
+                    - totalNumberItems
+                properties:
+                  pageNo:
+                    type: integer
+                    example: 1
+                  pageSize:
+                    type: integer
+                    example: 500
+                  totalNumberItems:
+                    type: integer
+                    example: 1
+                  items:
+                    type: array
+                    items:
+                      $ref: "#/components/schemas/EiffelEvent"
+        401:
+          description: Unauthorized
+          content: {}
+        403:
+          description: Forbidden
+          content: {}
+        404:
+          description: The requested events are not found
+          content: {}
+        500:
+          description: Internal server issue
+          content: {}
+  /events/{id}:
+    get:
+      tags:
+      - API
+      summary: To get single event information
+      operationId: getEventUsingGET
+      parameters:
+      - name: id
+        in: path
+        description: "Id of the event."
+        required: true
+        schema:
+          type: string
+      responses:
+        200:
+          description: Successfully retrieved the Event
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/EiffelEvent"
+        400:
+          description: Wrong type given for parameter
+          content: {}
+        401:
+          description: Unauthorized
+          content: {}
+        403:
+          description: Forbidden
+          content: {}
+        404:
+          description: The requested event is not found
+          content:
+            application/json:
+              schema:
+                type: object
+                properties:
+                  errorMsg:
+                    type: string
+                    example: "The requested event is not found"
+        500:
+          description: Internal server issue
+          content: {}
+  /search/{id}:
+    post:
+      tags:
+      - API
+      summary: To get upstream/downstream events for an event based on the searchParameters
+        passed
+      operationId: searchUsingPOST
+      parameters:
+      - name: id
+        in: path
+        description: "Id of the event."
+        required: true
+        schema:
+          type: string
+      - name: limit
+        in: query
+        description: >
+          Determines the maximum amount of events to be fetched.
+          Use `-1` for unlimited.
+        schema:
+          type: integer
+          format: int32
+      - name: levels
+        in: query
+        description: >
+          Determines the maximum amount of levels to search.
+          Use `-1` for unlimited
```

Comment:
I'm not sure it's a great idea to be able to ask for unlimited amounts of data. I guess an implementation can impose its own limit (i.e. cap the response), but if that limit is reached clients have no way of knowing.

Reply:
hmm, should we put a cap on the amount of events or should the implementation send back a fault describing it reached the maximum amount of events?

Reply:
Silent capping seems dangerous. Better to reject all requests with a too large `limit` or `levels` value with a 400 error. The actual limit should be implementation-dependent.

Reply:
I was thinking of specifying a cap in the api description e.g. "Valid range 1-1000" but I'll add a error code instead as you suggested when I get some time

Reply:
Fixed in [c2da26f](https://github.com/eiffel-community/eiffel-event-repository/pull/12/commits/c2da26f9974f462a0f2f68f76e1361fc5caa1183)

Reply:
I think you're misinterpreting what 413 is supposed to mean. AFAICT the "Payload too large" response references the payload sent by the client (in e.g. a POST request), not the payload that the server _would have_ sent. If the client requests too large of a response I think that's arguably a bad request, i.e. a 400 error. That's how Elasticsearch behaves in the same situation.

Reply:
I don't like using 400 as my interpretation is that is used for mostly syntax errors and such. A https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422 might work in this case but maybe its better with 500 error instead as it is a limitation on the server side.

Reply:
No, 500 is wrong. A client that requests too much data is clearly not an "unexpected condition [in the server] that prevented it from fulfilling the request". I'm fine with either 400 or 422. I found the discussion at https://stackoverflow.com/questions/16133923/400-vs-422-response-to-post-of-data to be pretty good and it simply isn't well-defined when to use 400 and when to use 422.

Reply:
Changed to 422 in [09ebfd8](https://github.com/eiffel-community/eiffel-event-repository/pull/12/commits/09ebfd8d6119e0310cc17a06d4fba3721f28b286)

Reply:
-1 is still listed as an allowed value (with no default) for `limit` and `levels`.

Reply:
Trying to understand: Should a user always have to specify a limit of events or should they skip sending the parameter to get "all" events (assuming the amount of events do not break the server capacity)? For us to be compliant with our current implementation we need to send -1 for "max" possible values.

Reply:
Having to pass -1 to get the max number of results allowed by the server sounds like a good idea. That way we can keep a sane default if the parameter is omitted (e.g. a few hundred events) yet allow a much higher limit that people can request if they want to (and they don't have to know what the limit is). We just have to document the meaning of -1.

Reply:
In ... I have:
* Described what `-1` stands for
* Added default values